paging/src/main/java/com/halcyonmobile/android/paging/PagingStateMachine.kt

```diff
@@ -23,21 +23,25 @@ class PagingStateMachine<T>(
     private val requestElements: suspend (numberOfElements: Int, isFetch: Boolean) -> Answer<T>
 ) {
 
-    private var cancelCurrentRequest: Job = Job()
+    private var currentRequest: Job = Job()
     private val stream = ConflatedBroadcastChannel<InternalState<T>>()
 
     val pagedDataStream: Flow<PagedResult<T>> = stream
-        .apply { offer(InternalState.ShowLoadingInitial(flowOf(emptyList()), 0, false, false)) } // todo refreshing
+        .apply { offer(InternalState.ShowLoadingInitial(flowOf(emptyList()), 0, false)) }
         .asFlow()
         .flowOn(dispatcher)
         .onEach { onInternalState(it) }
         .mapNotNull { internalStateToDataWithPagedState(it) }
         .flatMapLatest { it }
 
     fun fetch() {
-        cancelCurrentRequest.cancel()
-        cancelCurrentRequest = Job()
-        stream.offer(InternalState.ShowLoadingInitial(flowOf(emptyList()), 0, true, false))
+        (stream.valueOrNull as? InternalState.DoLoading)?.let {
+            // If it's already fetching or refreshing, do nothing
+            if (it.fetch) return
+        }
+        currentRequest.cancel()
+        currentRequest = Job()
+        stream.offer(InternalState.ShowLoadingInitial(flowOf(emptyList()), 0, true))
```

Comment:
So here is where your empty list is coming from. if you wish to have the same data/ datastream while refreshing you should provide here the previous flow, (or empty if it doesn't exists)